### PR TITLE
output plugin: syslog

### PR DESCRIPTION
**Pull request by javicrespo**
### Required for all PRs:
Concerns #5495 , I work with @MaxDeg
- [x] Signed [CLA](https://influxdata.com/community/cla/). 
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

### CODE REVIEW

File: plugins/outputs/syslog/README.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
+# Syslog Output Plugin
+
+The syslog output plugin sends syslog messages transmitted over
+[UDP](https://tools.ietf.org/html/rfc5426) or
+[TCP](https://tools.ietf.org/html/rfc6587) or
+[TLS](https://tools.ietf.org/html/rfc5425), with or without the octet counting framing.
+
+Syslog messages areformatted according to
```
Comment (leodido) [inline plugins/outputs/syslog/README.md]: Space missing here :)

File: plugins/outputs/syslog/README.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+## or the non-transparent framing technique (RFC6587#section-3.4.2).
+## Must be one of "octect-counting", "non-transparent".
+# framing = "octet-counting"
+
+## The trailer to be expected in case of non-trasparent framing (default = "LF").
+## Must be one of "LF", or "NUL".
+# trailer = "LF"
+
+### SD-PARAMs settings
+### A syslog message can contain multiple parameters and multiple identifiers within structured data section
```
Comment (leodido) [inline plugins/outputs/syslog/README.md]: I would rephrase this into

```suggestion
### A syslog message can contain multiple structured data sections.
### Thus multiple identifiers (one per section) and multiple parameters (multiple pairs per SD SECTION).
```

File: plugins/outputs/syslog/README.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+## Default sdid used for for fields that don't contain a prefix defined in the explict sdids setting below
+## If no default is specified, no SD-PARAMs will be used for unrecognised field.
+#default_sdid = "default@32473"
+
+##List of explicit prefixes to extract from fields and use as the SDID, if they match (see above example for more details):
+#sdids = ["foo@123", "bar@456"]
+###
+
+# Default PRI value (RFC5424#section-6.2.1) If no metric Field with key "PRI" is defined, this default value is used.
+default_priority = 0
```
Comment (leodido) [inline plugins/outputs/syslog/README.md]: Not sure we want the default to be `0` - ie., a kernel message (facility=0) with a severity of emergency (severity=0).

WDYT @danielnelson ?

File: plugins/outputs/syslog/README.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+### sdids = ["foo@123", "bar@456"]
+### input => xyzzy,x=y foo@123_value=42,bar@456_value2=84,something_else=1
+### output (structured data only) => [foo@123 value=42][bar@456 value2=84][default@32473 something_else=1 x=y]
+
+## SD-PARAMs separator between the sdid and field key (default = "_") 
+sdparam_separator = "_"
+
+## Default sdid used for for fields that don't contain a prefix defined in the explict sdids setting below
+## If no default is specified, no SD-PARAMs will be used for unrecognised field.
+#default_sdid = "default@32473"
```
Comment (leodido) [inline plugins/outputs/syslog/README.md]:
```suggestion
# default_sdid = "default@32473"
```

File: plugins/outputs/syslog/README.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+##List of explicit prefixes to extract from fields and use as the SDID, if they match (see above example for more details):
+#sdids = ["foo@123", "bar@456"]
```
Comment (leodido) [inline plugins/outputs/syslog/README.md]:
```suggestion
# sdids = ["foo@123", "bar@456"]
```

File: plugins/outputs/syslog/framing.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
+package syslog
```
Comment (leodido) [inline plugins/outputs/syslog/framing.go]: This seems a copy of https://github.com/influxdata/telegraf/blob/master/plugins/inputs/syslog/framing.go.

I personally try to avoid code duplication. So probably would be better to use the `"input/syslog/Framing"` type directly.

And BTW if we do not want plugins belonging to different categorization (inputs vs outputs) to depend on each other pieces we can factor our common pieces.

WDYT @danielnelson ?

File: plugins/outputs/syslog/framing_test.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
+package syslog
```
Comment (leodido) [inline plugins/outputs/syslog/framing_test.go]: Idem as above.

File: plugins/outputs/syslog/syslog.go [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+### sdids = ["foo@123", "bar@456"]
+### input => xyzzy,x=y foo@123_value=42,bar@456_value2=84,something_else=1
+### output (structured data only) => [foo@123 value=42][bar@456 value2=84][default@32473 something_else=1 x=y]
+
+## SD-PARAMs separator between the sdid and field key (default = "_") 
+sdparam_separator = "_"
+
+## Default sdid used for for fields that don't contain a prefix defined in the explict sdids setting below
+## If no default is specified, no SD-PARAMs will be used for unrecognised field.
+#default_sdid = "default@32473"
```
Comment (leodido) [inline plugins/outputs/syslog/syslog.go]:
```suggestion
# default_sdid = "default@32473"
```

File: plugins/outputs/syslog/syslog.go [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+
+##List of explicit prefixes to extract from fields and use as the SDID, if they match (see above example for more details):
+#sdids = ["foo@123", "bar@456"]
```
Comment (leodido) [inline plugins/outputs/syslog/syslog.go]:
```suggestion
# sdids = ["foo@123", "bar@456"]
```

File: plugins/outputs/syslog/framing.go (hunk 6 above)
Comment (javicrespo) [reply plugins/outputs/syslog/framing.go]: I hesitated to extract it from the input plugin and include it somewhere within telegraf/internal/ but I didn't see anything plugin specific in there, so I went with the copy&paste (I'm not a fan either :) )

Comment (leodido) [reply plugins/outputs/syslog/framing.go]: Let's wait for Daniel :)

P.S.: needs some time to do a squash of the commit github has done for me :')

**Comment by leodido** [comment]
Sorry @javicrespo I did a mistake during force-pushing on your fork's master.

I pushed your code into the [javicrespo-syslog-output branch](https://github.com/influxdata/telegraf/tree/javicrespo-syslog-output), could you please re-open the PR from that branch? Thanks and sorry again for the annoyance.
